urlb534bf3fca_akademi4eg_aka_uvc_tgz-pJ8-net_dtkanov_blocks_core_AppCoreJ8/src/net/dtkanov/blocks/circuit/high_level/derived/ControlUnit.java

```diff
@@ -0,0 +1,728 @@
+		is_mem_instr.connectSrc(comb_st_1, 0, 1);
+		///////////////////////////////////////////////////////////////////////
+		pc_ctrl = new PCRegController(PC);
+		// lowest bit of opcode = 0 => conditional jump
+		is_not_jump = new ORNode(alu_ctrl, 4, alu_ctrl, 5);
+		NOTNode neg_is_cond = new NOTNode();
+		neg_is_cond.connectSrc(opNOPs[0], 0, 0);// true when conditional jump
+		AllNode is_pchl = new AllNode(BITNESS);
+		NOTNode rev1_bit = new NOTNode(opNOPs[1], 0);
+		NOTNode rev2_bit = new NOTNode(opNOPs[2], 0);
+		NOTNode rev4_bit = new NOTNode(opNOPs[4], 0);
+		is_pchl.connectSrc(opNOPs[0], 0, 0);
+		is_pchl.connectSrc(rev1_bit, 0, 1);
+		is_pchl.connectSrc(rev2_bit, 0, 2);
+		is_pchl.connectSrc(opNOPs[3], 0, 3);
+		is_pchl.connectSrc(rev4_bit, 0, 4);
+		is_pchl.connectSrc(opNOPs[5], 0, 5);
+		is_pchl.connectSrc(opNOPs[6], 0, 6);
+		is_pchl.connectSrc(opNOPs[7], 0, 7);
+		ANDNode skip_jump = new ANDNode(new ANDNode(neg_is_cond, 0,
+													new NOTNode(is_pchl, 0), 0), 0,
+										not_jump_cond, 0);
+		Mux jumper = new Mux();
+		jumper.connectSrc(is_not_jump, 0, 2);
+		jumper.connectSrc(alu_ctrl, 4, 0);
+		jumper.connectSrc(skip_jump, 0, 1);
+		pc_ctrl.connectSrc(jumper, 0, BITNESS*2);
+		jumper = new Mux();
+		jumper.connectSrc(is_not_jump, 0, 2);
+		jumper.connectSrc(alu_ctrl, 5, 0);
+		jumper.connectSrc(skip_jump, 0, 1);
+		pc_ctrl.connectSrc(jumper, 0, BITNESS*2 + 1);
+		pc_ctrl.connectSrc(clock, 0, BITNESS*2 + 2);
+		MultiMux pc_src_low = new MultiMux(BITNESS);
+		pc_src_low.connectSrc(is_pchl, 0, 2*BITNESS);
+		MultiMux pc_src_high = new MultiMux(BITNESS);
+		pc_src_high.connectSrc(is_pchl, 0, 2*BITNESS);
+		for (int i = 0; i < BITNESS; i++) {
+			pc_src_low.connectSrc(L, i, i);
+			pc_src_low.connectSrc(inNOPs_A[i], 0, i+BITNESS);
+			pc_src_high.connectSrc(H, i, i);
+			pc_src_high.connectSrc(inNOPs_B[i], 0, i+BITNESS);
+			pc_ctrl.connectSrc(pc_src_low, i, i);
+			pc_ctrl.connectSrc(pc_src_high, i, i + BITNESS);
+		}
+		///////////////////////////////////////////////////////////////////////
+		is_alu_on = new NOPNode();
+		is_alu_on.connectSrc(alu_ctrl, 6, 0);
+		///////////////////////////////////////////////////////////////////////
+		
+		initOutputToRegisters();
+		initInputFromRegisters();
+	}
+	
+	private void initJumps() {
+		Mux flag_sel[] = new Mux[3];
+		flag_sel[0] = new Mux();
+		flag_sel[0].connectSrc(F, S_FLAG, 0);
+		flag_sel[0].connectSrc(F, C_FLAG, 1);
+		flag_sel[0].connectSrc(opNOPs[5], 0, 2);
+		flag_sel[1] = new Mux();
+		flag_sel[1].connectSrc(F, P_FLAG, 0);
+		flag_sel[1].connectSrc(F, Z_FLAG, 1);
+		flag_sel[1].connectSrc(opNOPs[5], 0, 2);
+		flag_sel[2] = new Mux();
+		flag_sel[2].connectSrc(flag_sel[0], 0, 0);
+		flag_sel[2].connectSrc(flag_sel[1], 0, 1);
+		flag_sel[2].connectSrc(opNOPs[4], 0, 2);
+		not_jump_cond = new XORNode();
+		not_jump_cond.connectSrc(flag_sel[2], 0, 0);
+		not_jump_cond.connectSrc(opNOPs[3], 0, 1);
+	}
+	
+	private void initInputFromRegisters() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 79 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 344-443 into a separate method.